targets/tofino/base/expression_resolver.cpp

```diff
@@ -35,6 +36,59 @@ auto ReturnDummyImpl = [](const ExternMethodImpls::ExternInfo &externInfo) {
     return returnVar;
 };
 
+const IR::Expression *RegisterActionExecuteWithDummyValue(
+    const ExternMethodImpls::ExternInfo &externInfo) {
+    auto &state = externInfo.state;
+    const auto *actionDecl =
+        state.findDecl(&externInfo.externObjectRef)->checkedTo<IR::Declaration_Instance>();
+    const auto *actionType = actionDecl->type->checkedTo<IR::Type_Specialized>();
+    const IR::Type *TType;
+    const IR::Type *UType;
+    if (actionType->baseType->path->name.name == "RegisterAction") {
+        TType = actionType->arguments->at(0);
+        UType = actionType->arguments->at(2);
+    } else if (actionType->baseType->path->name.name == "DirectRegisterAction") {
+        TType = actionType->arguments->at(0);
+        UType = actionType->arguments->at(1);
+    } else {
+        BUG("Unsupported action type %1%", actionType->baseType->path->name.name);
+    }
+    const auto *applyDecl = actionDecl->initializer->components.at(0)->checkedTo<IR::Function>();
+    const auto *applyParameters = applyDecl->type->parameters;
+
+    const auto &valueName = applyParameters->parameters.at(0)->name.name;
+    auto valueLabel = externInfo.externObjectRef.path->toString() + "_" + externInfo.methodName +
+                      "_" + std::to_string(externInfo.originalCall.clone_id) + "_apply_" +
+                      valueName;
+    /// TODO: currently we create symbolic expression as the value, but we should model the
+    /// values stored in register.
+    const auto *valueExpr = state.createSymbolicExpression(TType, valueLabel);
+
+    auto returnLabel = externInfo.externObjectRef.path->toString() + "_" + externInfo.methodName +
+                       "_" + std::to_string(externInfo.originalCall.clone_id);
+    const auto *returnVar = ToolsVariables::getSymbolicVariable(UType, returnLabel);
+
+    IR::Vector<IR::Argument> arguments = {new IR::Argument(valueExpr), new IR::Argument(returnVar)};
+
+    for (size_t argIdx = 0; argIdx < applyParameters->size(); ++argIdx) {
```

Comment:
Since I am returning the result from state([here](https://github.com/fruffy/flay/blob/zhanghan/register-action/targets/tofino/base/expression_resolver.cpp#L481)), do we still need to model copy-in/out semantics? (Not sure if I fully understand the copy-in/out semantics).

Reply:
`out` resets a parameter to an undefined value or target-specific value (0 for BMv2, for example). The parameter is effectively write-only.  `in` specifies that the parameter is read-only. `inout` specifies that the parameter can be read and written. 

`out` is special because it deinitializes anything that gets passed in. We have to take that in consideration when calling execute.

Reply:
Ah I see. Thanks for pointing out.